crates/fluvio-protocol/src/core/bytebuf.rs

```diff
@@ -0,0 +1,301 @@
+use std::io::Error;
+use std::io::ErrorKind;
+
+use bytes::buf::UninitSlice;
+use bytes::{Buf, Bytes, BufMut};
+
+use crate::{Encoder, Decoder, Version};
+
+/// Represnts a SmartModule WASM File bytes.
+///
+/// Provides a `Encoder` and `Decoder` implementation optimized for WASM files
+/// used in SmartModules.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct ByteBuf {
+    inner: Vec<u8>,
+}
+
+impl ByteBuf {
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn inner(&self) -> Vec<u8> {
+        self.inner.clone()
+    }
+}
+
+impl From<Vec<u8>> for ByteBuf {
+    fn from(bytes: Vec<u8>) -> Self {
+        ByteBuf { inner: bytes }
+    }
+}
+
+impl Decoder for ByteBuf {
+    fn decode<T>(&mut self, src: &mut T, version: Version) -> Result<(), Error>
+    where
+        T: Buf,
+    {
+        let mut len: i32 = 0;
+        len.decode(src, version)?;
+
+        if len < 1 {
+            return Ok(());
+        }
+
+        let mut buf = src.take(len as usize);
+        self.inner
+            .extend_from_slice(&buf.copy_to_bytes(len as usize));
+
+        Ok(())
+    }
+}
+
+impl Encoder for ByteBuf {
+    fn write_size(&self, version: Version) -> usize {
+        self.inner
+            .iter()
+            .fold(4, |sum, val| sum + val.write_size(version))
+    }
+
+    fn encode<T>(&self, dest: &mut T, version: Version) -> Result<(), Error>
+    where
+        T: BufMut,
+    {
+        let remaining = dest.remaining_mut();
+        let expected = self.write_size(version);
+
+        if remaining < expected {
+            return Err(Error::new(
+                ErrorKind::UnexpectedEof,
+                format!(
+                    "Not enough capacity for ByteBuf. Expected: {}, Remaining: {}",
+                    expected, remaining
+                ),
+            ));
+        }
+
+        dest.put_u32(self.inner.len() as u32);
+        dest.put_slice(self.inner.as_slice());
+
+        Ok(())
+    }
+
+    fn as_bytes(&self, _version: Version) -> Result<Bytes, Error> {
+        Ok(Bytes::copy_from_slice(self.inner.as_slice()))
+    }
+}
+
+unsafe impl BufMut for ByteBuf {
```

Comment:
Is there way to do without unsafe implementation?

Reply:
I agree, it's always better to avoid `unsafe`. I searched for alternatives but theres no derive macro that could apply this for us.

This is required in order to call decode on `ByteBuf`, and given that `BufMut` is already implemented for `Vec<u8>` by the crate authors, I took advantage of such implementation by just calling the respective trait functuons for the `Vec<u8>` in the `inner` prop.

Reply:
Can you try Deref?

Reply:
Of course!

Reply:
I did try by removing the implementation of `ByteBuf`, but it complains about missing impl even when `Deref` is implemented already.